torch/distributions/distribution.py

```diff
@@ -1,15 +1,25 @@
 from torch.autograd import Variable
+import warnings
 
 
 class Distribution(object):
     r"""
     Distribution is the abstract base class for probability distributions.
     """
+    has_rsample = False
 
-    def sample(self):
+    def sample(self, sample_shape=()):
         """
-        Generates a single sample or single batch of samples if the distribution
-        parameters are batched.
+        Generates a sample_shape shaped sample or sample_shape shaped batch of samples if the distribution
+        parameters are batched. Currently only supports len(sample_shape)<2.
+        """
+        z = self.rsample(sample_shape)
+        return z.detach() if hasattr(z, 'detach') else z
```

Comment:
why is this `hasattr` needed?

Reply:
Distributions sample Tensors if parameters are Tensors or Variables if parameters are Variables. So `z if is_tensor(z) else z.detach()` should also work.